Update conceptual documentation for v3 GA (#287)
MicrosoftDocs/PowerShell-Docs-DSC: dsc/docs-conceptual/dsc-3.0/get-started/index.md

DSC resources to query, audit, and set the state of a specific component on a system. You can use
DSC configuration documents to describe how a system should be configured using one or more
resources. This document describes how you can discover the resources available on a machine,
invoke those resources directly, and manage a configuration.

## Prerequisites

- [Install DSC][01] version `3.0.0` on a Windows machine.
- A terminal emulator, like [Windows Terminal][02].

## Discover resources

You can use the `dsc resource list` command to enumerate the available DSC resources on a machine.
DSC discovers resources by searching the folders in your `PATH` environment variable for files that
have any of the following suffixes:

- `.dsc.resource.json`
- `.dsc.resource.yaml`
- `.dsc.resource.yml`

Files with the `.dsc.resource.<extension>` suffix are DSC resource manifests. They describe both
the settings they enable you to manage and how DSC can invoke them as a command.

Open a terminal and run the following command:

```powershell
dsc resource list
```

DSC outputs a table showing the available resources and summary information about each resource.

```Output

> Type                                        Kind      Version  Capabilities  RequireAdapter  Description                                                                
> ------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> Microsoft.DSC.Debug/Echo                    Resource  1.0.0    gs--t---
> Microsoft.DSC.Transitional/RunCommandOnSet  Resource  0.1.0    gs------                      Takes a single-command line to execute on DSC set operation
> Microsoft.DSC/Assertion                     Group     0.1.0    gs--t---                      `test` will be invoked for all resources in the supplied configuration.    
> Microsoft.DSC/Group                         Group     0.1.0    gs--t---                      All resources in the supplied configuration is treated as a group.
> Microsoft.DSC/Include                       Importer  0.1.0    gs--t---                      Allows including a configuration file with optional parameter file.        

Reviewer: Trim trailing blanks
```suggestion
Type                                        Kind      Version  Capabilities  RequireAdapter  Description
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Microsoft.DSC.Debug/Echo                    Resource  1.0.0    gs--t---
Microsoft.DSC.Transitional/RunCommandOnSet  Resource  0.1.0    gs------                      Takes a single-command line to execute on DSC set operation
Microsoft.DSC/Assertion                     Group     0.1.0    gs--t---                      `test` will be invoked for all resources in the supplied configuration.
Microsoft.DSC/Group                         Group     0.1.0    gs--t---                      All resources in the supplied configuration is treated as a group.
Microsoft.DSC/Include                       Importer  0.1.0    gs--t---                      Allows including a configuration file with optional parameter file.
```

Microsoft.DSC/PowerShell                    Adapter   0.1.0    gs--t-e-                      Resource adapter to classic DSC Powershell resources.
Microsoft.Windows/RebootPending             Resource  0.1.0    g-------                      Returns info about pending reboot.
Microsoft.Windows/Registry                  Resource  0.1.0    gs-w-d--                      Manage Windows Registry keys and values
Microsoft.Windows/WMI                       Adapter   0.1.0    g-------                      Resource adapter to WMI resources.
Microsoft.Windows/WindowsPowerShell         Adapter   0.1.0    gs--t---                      Resource adapter to classic DSC Powershell resources in Windows PowerShell.
Microsoft/OSInfo                            Resource  0.1.0    g-----e-                      Returns information about the operating system.
Microsoft/Process                           Resource  0.1.0    gs--t-e-                      Returns information about running processes.
```

Together, the columns describe each resource:

> - The **Type** column defines the fully qualified type name for a resource. This identifies the
>   resource on a system and in a configuration document.

Reviewer: ```suggestion
- The **Type** column defines the fully qualified type name for a resource. This name identifies the
  resource on a system and in a configuration document.
```

  - `x` indicates that the resource has the `setHandlesExist` capability.
  - `w` indicates that the resource has the `whatIf` capability.
  - `t` indicates that the resource has the `test` capability.
  - `d` indicates that the resource has the `delete` capability.
  - `e` indicates that the resource has the `export` capability.
  - `r` indicates that the resource has the `resolve` capability.

- The **RequireAdapter** column indicates which adapter resource, if any, the resource requires.
- The **Description** column defines a synopsis for the resource.

By default, the `dsc resource list` command only returns command resources, which always have a
resource manifest. You can use the `--adapter` option to return the resources for one or more
adapted resources.

Run the following command to return the available adapted resources instead of command resources:

```powershell
$adaptedResources = dsc resource list --adapter * | ConvertFrom-Json

$adaptedResources |
  Group-Object -NoElement -Property requireAdapter |
  Format-Table -AutoSize
```

```Output
Count Name
----- ----
   27 Microsoft.Windows/WindowsPowerShell
  961 Microsoft.Windows/WMI
```

> The first command saves the output of `dsc resource list` to the `$resources` variable as an array
> of PowerShell objects. When you first ran the `dsc resource list` command, the output in the
> terminal was a table view. By default, when DSC detects that its output is being redirected to a
> file, variable, or another command in the pipeline, it emits JSON representing the output.
> Converting the JSON into a PowerShell object with the `ConvertFrom-Json` cmdlet enables you to
> treat the output like any other PowerShell object.
> 
> The second command uses the `Group-Object` cmdlet to summarize the available adapted resources,
> grouped by the adapter that they require. The specific count for adapted resources depends on the
> machine you're using and whether you've installed any PowerShell modules that export any PSDSC
> resources.

Reviewer: ```suggestion
The output of `dsc resource list` is saved to the `$resources` variable as an array of PowerShell
objects. When you first ran the `dsc resource list` command, the output in the terminal was a table
view. By default, when DSC detects output redirection, it formats the output as JSON. Converting the
JSON into a PowerShell object with the `ConvertFrom-Json` cmdlet enables you to treat the output
like any other PowerShell object.

The `Group-Object` cmdlet summarizes the available adapted resources, grouped by the adapter that
they require. The specific count for adapted resources depends on the machine you're using and the
installed PowerShell modules that export any PSDSC resources.
```

- The `set` capability indicates that you can invoke the **Set** operation to modify the state of
  an instance.
- The `whatIf` capability indicates that you can invoke the **Set** operation in what-if mode to
  see how the resource would change the instance without actually modifying the system.
- The `delete` capability indicates that you can invoke the **Delete** operation to remove an
  instance from the system.

### Get the current state of a resource

You can use the `dsc resource get` command to retrieve the current state of an instance. Run the
following command to define the data that represents a registry value:

```powershell
$instance = @{
  keyPath   = 'HKLM\Software\Microsoft\Windows NT\CurrentVersion'
  valueName = 'SystemRoot'
} | ConvertTo-Json
```

For this example:

- The `keyPath` property to indicate the path to the registry key that should contain the value you
  want to manage. `keyPath` is a _required_ property for the `Registry` resource - you always need
  to specify it.
- The `valueName` property identifies which registry value to manage for the registry key.
- The `_exist` canonical property indicates whether the registry value should exist. In this

>   example, it's set to `true`, indicating the instance should exist.

Reviewer: ```suggestion
  example, the value is `true`, which indicates that the instance should exist.
```
